cleverhans/loss.py

```diff
@@ -329,3 +334,100 @@ def fprop(self, x, y, **kwargs):
                   "MixUp. LossFeaturePairing may be removed "
                   "on or after 2019-03-06.")
     return loss
+
+
+class SNNLCrossEntropy(CrossEntropy):
+  EPSILON = 0.00001
+
+  def __init__(self,
+               model,
+               temperature=100.,
+               layer_names=None,
+               factor=-10.,
+               optimize_temperature=True,
+               cos_distance=False):
+    CrossEntropy.__init__(self, model, smoothing=0.)
+    self.temperature = temperature
+    self.factor = factor
+    self.optimize_temperature = optimize_temperature
+    self.cos_distance = cos_distance
+    self.layer_names = layer_names
+    if not layer_names:
+      # omit the final classification layer
+      self.layer_names = model.get_layer_names()[:-1]
+
+  @staticmethod
+  def pairwise_euclid_distance(A, B):
+    batchA = tf.shape(A)[0]
+    batchB = tf.shape(B)[0]
+
+    sqr_norm_A = tf.reshape(tf.reduce_sum(tf.pow(A, 2), 1), [1, batchA])
+    sqr_norm_B = tf.reshape(tf.reduce_sum(tf.pow(B, 2), 1), [batchB, 1])
+    inner_prod = tf.matmul(B, A, transpose_b=True)
+
+    tile_1 = tf.tile(sqr_norm_A, [batchB, 1])
+    tile_2 = tf.tile(sqr_norm_B, [1, batchA])
+    return (tile_1 + tile_2 - 2 * inner_prod)
+
+  @staticmethod
+  def pairwise_cos_distance(A, B):
+    normalized_A = tf.nn.l2_normalize(A, dim=1)
+    normalized_B = tf.nn.l2_normalize(B, dim=1)
+    prod = tf.matmul(normalized_A, normalized_B, adjoint_b=True)
+    return 1 - prod
+
+  @staticmethod
+  def fits(A, B, temp, cos_distance):
+    if cos_distance:
+      distance_matrix = SNNLCrossEntropy.pairwise_cos_distance(A, B)
+    else:
+      distance_matrix = SNNLCrossEntropy.pairwise_euclid_distance(A, B)
+    return tf.exp(-(distance_matrix / temp))
+
+  @staticmethod
+  def pick_probability(x, temp, cos_distance):
```

Comment:
I'll quit writing "please add docstring", but please add docstrings to all the new methods / functions in the PR